### DISCUSSION TITLE
[FEATURE REQUEST] add autoreboot boolean parameter to win_wua.list install=True, and win_wua.uptodate

### DISCUSSION THREAD

**Discussion by ewhitesides** [saltstack/salt#67323] [Feature Requests]
it would be great if you could incorporate an autoreboot switch, so that when updates are complete and one or more updates require a reboot, the machine will autoreboot.  

this would be easier than having to issue a 'get reboot status', check that all machines are returning true, then sending a reboot command.

**Comment by xeacott** [comment]
Our next release Sodium does have a `pending_reboot` grain, which can be utilized for this. We may add the ability to the functions themselves, so I won't close this ticket, but we aren't considering it right now.

That grain is in master right now so you can give it a shot.

**Comment by ewhitesides** [comment]
The general idea is that it would be less moving parts to have the install function check that grain for me.

What if I were trying to check 1000 servers, every 5 mins for that grain? What if the salt-master was busy, so I would then need to call it with an async switch, so then my script would then need to parse json on a job lookup? You get the idea, the extra complexity starts snowballing.

**Comment by xeacott** [comment]
Sure, and we're trying to weigh all these considerations especially because with an auto_reboot feature, there could be just as many instances where someone _doesn't_ want to auto-reboot because it will take 1000 machines offline during production time and various intervals.

That's why we went in the direction of the grain right now and to give users the ability to monitor their own scheduling based on checking that grain. 

On the one hand, it does introduce potential overhead if you need to hit a server every 5 minutes to check which of your 1000 machines need to be restarted. When that situation arises, it's usually a good idea to do A/B patching, where only half your machines are patched on week 1 and then the rest on week 2. And it depends how regularly you want to check for updates, alluding to the almighty "patch tuesdays".

On the other hand, having autoreboot does give certain users a much easier time using the methods and not need to check and potentially re-calculate the grains. 

Not fighting you on this or trying to give push-back... but trying to think it through.

**Comment by ewhitesides** [comment]
I would leave autoreboot as default to False, it's not something everyone will want.

**Comment by xeacott** [comment]
I'll get a pull request together for this and we can see what it looks like and decide over it. 😄

**Comment by xeacott** [comment]
Because personally I would like to see this.

**Comment by xeacott** [comment]
I decided not to make a PR for this yet ... simply because this is an opportunity to refactor and condense a lot of code. There's various pending_reboot / needs_reboot functions that exist that need to be moved, there's utility functions that are calling state functions, etc. 

so instead, on my fork i have a branch called `wua_auto_reboot`. if you'd like to see the diff, go for it.
